File: app/src/main/java/org/jellyfin/androidtv/ui/ExpandableTextView.java
package org.jellyfin.androidtv.ui;

import android.content.Context;
import android.content.Intent;
import android.graphics.Canvas;
import android.graphics.Rect;
import android.text.Layout;
import android.util.AttributeSet;
import android.view.SoundEffectConstants;
import android.view.View;

import org.jellyfin.androidtv.R;
import org.jellyfin.androidtv.TvApp;
import org.jellyfin.androidtv.details.ExpandedTextActivity;

import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.app.ActivityOptionsCompat;

/**
 * Created by spam on 9/30/2016.
 */

public class ExpandableTextView extends AppCompatTextView {

    private boolean textChanged;
    private boolean isElipsized;

    public ExpandableTextView(Context context) {
        super(context);
        init();
    }

    public ExpandableTextView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        final View us = this;
        setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                playSoundEffect(SoundEffectConstants.CLICK);
                Intent expanded = new Intent(TvApp.getApplication(), ExpandedTextActivity.class);
                expanded.putExtra("text", getText().toString());
                ActivityOptionsCompat options = ActivityOptionsCompat.
                        makeSceneTransitionAnimation(TvApp.getApplication().getCurrentActivity(), us, "summary");
                TvApp.getApplication().getCurrentActivity().startActivity(expanded, options.toBundle());
            }
        });

    }

    @Override
    public void setText(CharSequence text, BufferType type) {
        super.setText(text, type);
        textChanged = true;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (textChanged) {
            textChanged = false;
            Layout l = getLayout();
            if ( l != null) {
                int lines = l.getLineCount();
                isElipsized = lines > 0 && l.getEllipsisCount(lines - 1) > 0;
                setFocusable(isElipsized);
            } else
                isElipsized = false;

        }
    }

    @Override
    protected void onFocusChanged(boolean focused, int direction, Rect previouslyFocusedRect) {
        super.onFocusChanged(focused, direction, previouslyFocusedRect);

        if (focused) {
            setBackgroundResource(R.drawable.txt_focus);
        } else {
            setBackgroundColor(0);
        }
    }

    public boolean isElipsized() {
        return isElipsized;
    }
}

File: app/src/main/java/org/jellyfin/androidtv/ui/ExpandableTextView.kt
package org.jellyfin.androidtv.ui

import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.PopupWindow
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.ContextCompat
import org.jellyfin.androidtv.R
import org.jellyfin.androidtv.databinding.PopupExpandableTextViewBinding

class ExpandableTextView @JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null,
	defStyleAttr: Int = 0,
) : AppCompatTextView(context, attrs, defStyleAttr) {
	private val popupContentBinding = PopupExpandableTextViewBinding.inflate(
		LayoutInflater.from(context),
		null,
		false
	)
	private val popup = PopupWindow(
		popupContentBinding.root,
		ViewGroup.LayoutParams.MATCH_PARENT,
		ViewGroup.LayoutParams.MATCH_PARENT,
		true
	).apply {
		animationStyle = R.style.WindowAnimation_Fade
	}

	init {
		background = ContextCompat.getDrawable(context, R.drawable.expanded_text)

		popupContentBinding.scrollContainer.setOnClickListener { popup.dismiss() }

		setOnClickListener {
			// Update text
			popupContentBinding.content.text = this@ExpandableTextView.text

			// Show popup
			popup.showAtLocation(rootView, Gravity.CENTER, 0, 0)
			popup.isOutsideTouchable = true

			// Focus on scroll view
			popupContentBinding.scrollContainer.requestFocus()
		}
	}

	override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
		super.onLayout(changed, left, top, right, bottom)

		// Only make focusable when text is ellipsized.
		val isEllipsized = !text.isNullOrBlank() && lineCount > 0 && layout.getEllipsisCount(lineCount - 1) > 0
		isFocusable = isEllipsized
		isClickable = isEllipsized
	}
}
